internal/grpc/commander/connection.go

```diff
@@ -93,19 +249,123 @@ func (c *connection) receive(ctx context.Context) error {
+		return
+	}
+
+	c.pendingConfig = config
+
+	c.logger.Info("Updating to latest config", "config generation", config.ID)
+
+	status := c.statusForID(ctx, config.ID)
+
+	select {
+	case <-ctx.Done():
+		c.logger.Error(ctx.Err(), "failed to update config")
+		return
+	case c.cmdExchanger.Out() <- createDownloadCommand(config.ID, c.nginxID, c.systemID):
+	}
+
+	now := time.Now()
+	c.logger.Info("Waiting for config status", "config generation", config.ID)
+
+	select {
+	case <-ctx.Done():
+		return
+	case s := <-status:
+		elapsedTime := time.Since(now)
+		c.logger.Info(
+			fmt.Sprintf("Config apply complete [%s]", s.status),
+			"message",
+			s.message,
+			"config generation",
+			config.ID,
+			"duration",
+			elapsedTime.String(),
+		)
+	}
+}
+
+// statusForID returns a channel that will receive a configApplyResponse when a final (
+// not pending) status is received for the given config ID.
+func (c *connection) statusForID(ctx context.Context, id string) <-chan configApplyResponse {
+	statusForID := make(chan configApplyResponse)
+
+	go func() {
+		defer close(statusForID)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case status := <-c.configApplyResponseCh:
+				// Not every status contains a correlation ID, so we only need to check it if it's not empty.
+				// This is a workaround for some inconsistencies in the way the agent reports config apply statuses.
+				if status.correlationID != "" && status.correlationID != id {
+					c.logger.Info("Config status is for wrong generation",
+						"actual config generation",
+						status.correlationID,
+						"expected config generation",
+						id,
+						"status",
+						status.status,
+						"message",
+						status.message,
+					)
+					continue
+				}
+
+				select {
+				case <-ctx.Done():
+					return
+				case statusForID <- status:
+					return
+				}
+			}
+		}
+	}()
+
+	return statusForID
+}
+
 func (c *connection) handleCommand(ctx context.Context, cmd *proto.Command) {
 	switch cmd.Data.(type) {
 	case *proto.Command_AgentConnectRequest:
-		c.handleAgentConnectRequest(ctx, cmd)
+		c.handleAgentConnectRequestCmd(ctx, cmd)
+	case *proto.Command_DataplaneStatus:
+		c.handleDataplaneStatus(ctx, cmd.GetDataplaneStatus())
+	case *proto.Command_NginxConfigResponse:
+		c.handleNginxConfigResponse(ctx, cmd.GetNginxConfigResponse())
 	default:
-		c.logger.Info("Ignoring command", "command data type", fmt.Sprintf("%T", cmd.Data))
+		c.logger.Info("Ignoring command", "data type", fmt.Sprintf("%T", cmd.Data))
 	}
 }
 
-func (c *connection) handleAgentConnectRequest(ctx context.Context, cmd *proto.Command) {
+func (c *connection) handleAgentConnectRequestCmd(ctx context.Context, cmd *proto.Command) {
 	req := cmd.GetAgentConnectRequest()
 
-	c.logger.Info("Received agent connect request", "message ID", cmd.GetMeta().GetMessageId())
+	c.logger.Info("Received agent connect request")
+
+	c.logger = c.logger.WithValues("podName", req.GetMeta().DisplayName)
```

Comment:
@pleshakov can you think of any reason adding a value to the logger during runtime would cause a problem? I _think_ this is ok, but I might be missing something.

Reply:
> perhaps we can refactor the connection so that when it starts, it always waits for the connect to finish, it then extracts the pod name, updates the logger and starts the connection?

Intriguing idea. I'll play around with it.